typed_sql/lib/src/adaptor/adaptor.dart

```diff
@@ -0,0 +1,330 @@
+  /// for details.
+  @visibleForTesting
+  factory DatabaseAdaptor.postgresTestDatabase({
+    String? host,
+    int? port,
+    String? database,
+    String? user,
+    String? password,
+  }) =>
+      futureDatabaseAdaptor(postgresTestingDatabaseAdaptor(
+        host: host,
+        port: port,
+        database: database,
+        user: user,
+        password: password,
+      ));
+
+  /// Wrap [adaptor] such that [close] calls [onClosed] instead of
+  /// `adaptor.close`.
+  ///
+  /// > [!NOTE]
+  /// > [adaptor] will not be closed when [close] is called on the returned
+  /// > [DatabaseAdaptor].
+  factory DatabaseAdaptor.withOnClose(
+    DatabaseAdaptor adaptor,
+    Future<void> Function({required bool force}) onClosed,
+  ) =>
+      withOnCloseDatabaseAdaptor(adaptor, onClosed);
+
+  /// Wrap [adaptor] such that [close] is a no-op!
+  ///
+  /// > [!NOTE]
+  /// > [adaptor] will not be closed when [close] is called on the returned
+  /// > [DatabaseAdaptor].
+  factory DatabaseAdaptor.withNopClose(DatabaseAdaptor adaptor) =>
+      withOnCloseDatabaseAdaptor(adaptor, ({required bool force}) async {});
+
+  /// Wrap [adaptor] such that all queries are written to [log].
+  ///
+  /// > [!WARNING]
+  /// > The format of log messages written are not covered by version
+  /// > compatibility guarantees.
+  ///
+  /// This is intended for debugging purposes, and pairs nicely with
+  /// [printOnFailure] from `package:test/test.dart`.
+  factory DatabaseAdaptor.withLogging(
+    DatabaseAdaptor adaptor,
+    void Function(String message) log,
+  ) =>
+      loggingAdaptor(adaptor, log);
+}
+
+/// Interface for executing database operations in a transaction.
+abstract base class DatabaseTransaction extends Executor {
+  /// {@macro Executor.query}
+  ///
+  /// The returned [Stream] must disregard back-pressure to avoid deadlocks in
+  /// the transaction. In practice this can be done by buffering all results in
+  /// a [List] and return a [Stream] wrapping said list.
+  @override
+  Stream<RowReader> query(String sql, List<Object?> params);
+
+  /// {@macro Executor.transact}
+  ///
+  /// Any call to [script], [query], [execute] or [transact] while a transaction
+  /// is ongoing shall throw [StateError], until the transaction has been
+  /// committed or rolled back.
+  @override
+  Future<T> transact<T>(Future<T> Function(DatabaseTransaction tx) fn);
+}
+
+/// Result of executing an SQL statement.
+final class QueryResult {
+  /// Number of rows affected by the SQL statement.
+  final int affectedRows;
+
+  QueryResult({
+    required this.affectedRows,
+  });
+}
+
+/// Interface for executing database operations.
+abstract final class Executor {
+  /// Execute an [sql] script that may contain multiple statements.
+  ///
+  /// Statements are typically separated by `;`.
+  ///
+  /// Throws [DatabaseException], if the execution fails.
+  Future<void> script(String sql);
```

Comment:
Not critical, but I expect that at some point you want to have a way of "preparing" statements, so you can run them multiple times against different arguments...